test/controllers/tasks_controller_test.rb

```diff
@@ -83,42 +79,193 @@
+
+      # success indicates rendering of page
+      assert_response :success
+    end
   end
   
-  # Unskip and complete these tests for Wave 3
+  # DONE - Unskip and complete these tests for Wave 3
   describe "edit" do
     it "can get the edit page for an existing task" do
-      skip
-      # Your code here
+      # Act
+      get edit_task_path(task.id)
+
+      # Assert
+      must_respond_with :success
     end
     
     it "will respond with redirect when attempting to edit a nonexistant task" do
-      skip
-      # Your code here
+      # Act
+      get edit_task_path(-1)
+
+      # Assert
+      must_respond_with :redirect
     end
   end
   
   # Uncomment and complete these tests for Wave 3
   describe "update" do
+    # Arrange
+    before do
+      Task.create(name:"edit this task name", description: "edit this description")
+    end
+    let (:edit_task_data) {
+      {
+        task: {name: "Task edited!",
+                description:"Description edited!"}
+      }
+    }
     # Note:  If there was a way to fail to save the changes to a task, that would be a great
-    #        thing to test.
+    #        thing to test - used validation and added a test for "create" as well!
     it "can update an existing task" do
-      # Your code here
+      # Arrange
+      id = Task.find_by(name:"edit this task name")[:id]
+      # Act-Assert
+      expect{
+        patch task_path(id), params: edit_task_data
+      }.wont_change "Task.count"
+
+      must_redirect_to task_path(id)
+
+      edited_task = Task.find_by(name: edit_task_data[:task][:name])
+      expect(edited_task.description).must_equal edit_task_data[:task][:description]
+      expect(edited_task.completed_at).must_equal edit_task_data[:task][:completed_at]
     end
     
     it "will redirect to the root page if given an invalid id" do
-      # Your code here
+      id = -1
+      # Act-Assert
+      expect {
+        patch task_path(id), params: edit_task_data
+      }.wont_change "Task.count"
+
+      must_redirect_to :root
+    end
+
+    it "will redirect to the root page if given an invalid id" do
+      id = -1
+      # Act-Assert
+      expect {
+        patch task_path(id), params: edit_task_data
+      }.wont_change "Task.count"
+
+      must_redirect_to :root
+    end
+
+    it "will render :edit if attempts to enter an empty task are made" do
+      # Arrange
+      id = Task.find_by(name:"edit this task name")[:id]
+      empty_task = {task: {name:nil, description:nil}}
+
+      # Act-Assert
+      expect {
+        patch task_path(id), params: empty_task
+      }.wont_change "Task.count"
+
+      # success indicates rendering of page
+      assert_response :success
     end
   end
   
-  # Complete these tests for Wave 4
+  # DONE - Complete these tests for Wave 4
   describe "destroy" do
-    # Your tests go here
-    
+    it "can destroy a model" do
+      # Arrange
+      delete_me = Task.new name: "Delete me", description: "Delete me"
```

Comment:
This syntax is correct, but it doesn't have the parentheses. To match the rest of your code style (which uses a lot of parens, namely around Task.new), it might be more consistent to have `Task.new(name: "", description:"")`